### ISSUE TITLE
Todo in order 3/14/2020

### ISSUE THREAD

**Issue by real-ian** [real-social-media/frontend#30]
# Photos
- [x] If you zoom in, then quickly back out, then zoom back In anytime, the native quality never downloads. Even if you zoom back in and stay zoomed in for 1 hour. 
- [x] When uploading multiple images at once, the previews mismatch the comments. When writing a comment, it uploads that comment with the wrong picture to create the post. 
- [x] tapping trending photos opens the wrong photo https://drive.google.com/file/d/1r7-qUJmFxpyu9Wqag4AHFVYV-RC3s9D-/view?usp=sharing
- [x] I see a lot of "64p" pictures in my news feed. They should have loading icons & be replaced with 4K quality. When zooming on them I see the 4K image is already downloaded, but I never see it in my news feed. Even after refreshing, it's still 64p. But zooming I see 4K & native. https://drive.google.com/file/d/1jkPdN-eiFKEWp66iZrVOAwifM6zFFrBb/view?usp=sharing
- [x] When zooming, I see native quality, but zooming back out, I still see 4K quality until I refresh the news feed. It should be native everywhere, after native quality is downloaded (due to a zoom) without requiring the news feed to be refreshed. Is there a way to refresh only that single picture. 
- [ ] When viewing my own profile (ian), I still see black, no pictures. Even after dragging down to refresh on the profile screen I still see black, logging out and back in does not solve the issue. Closing & re opening the app does not solve the issue. reinstalling the app does not solve the issue. https://drive.google.com/file/d/1eHNq-QpRjqru_E8ePtzhNs59gDE89mJE/view?usp=sharing 
- [x] When switching apps mid zoom - then coming back to REAL, REAL is frozen on the zoomed in image (forever). 
- [x] Add loading icons to thumbnails on user profile page & trending page. (any large image which is a thumbnail should have a loading icon on it, until the higher quality version finishes downloading). 
- [x] Pinch and Zoom images are blurry
- [x] Pinch and Zoom priority queue
- [x] Profile photos are blurry at first
- [x] Discover app speed
# Bugs
- [ ] Choose profile photo from gallery during signup is broken. It should open the camera library (iOS) 
- [x] hitting cancel on any photo during multiple photo upload (gallery) cancels entire upload, not just that image. 
- [ ] Can’t add an album to a post under edit post. It shows only “create new album” which also doesn’t work. No existing albums are shown 
- [x] Following says “followed” once open. It should say “Following”.
- [x] Resuming the app, I'm stuck on the search screen (frozen). Restarting the app fixed it. 
- [x] [to be verified] Fix reset password issues, getting hung (forever loading) on "Send Confirmation Email" to reset password. Also getting "Failed" after entering a new password on another account.
- [x] [to be verified] Creating album while uploading multiple posts closes & cancels post upload after tapping "create album" and entering an album name.
- [x] I can't login with google in the latest version. Clicking login with google just spins a wheel for 1-2 seconds, then nothing happens. Pressing login with google again does the same. https://drive.google.com/file/d/147G7H1HNg6eY51Z6gA0UgXCiyWLvYzVw/view?usp=sharing
- [x] when leaving the app, resuming I'm stuck on the search page (frozen forever). https://drive.google.com/file/d/17nN5t-RrWP0Le6vrWqXcsbRSUQ6lRfg4/view?usp=sharing
- [x] There should be loading icons on posts which are loading in the user profile. Currently I just see black squares until images appear / load. 
- [x] User profile & other users profiles can show all grey thumbnails. Refreshing (dragging down) does not fix the issue. Tapping a grey outline of a thumbnail does not open it. https://drive.google.com/file/d/1ll1VCnK2VVMZCUgw3DgVrjBmcgg61inh/view?usp=sharing
- [x] Images inside the share screen should be cached (the same image as shown in the news feed. https://drive.google.com/file/d/1lkYkOUO9KICa6-G3T6bQAcVvAcppj94z/view?usp=sharing
- [x] When downloading native quality image (user tried to pinch & zoom), the native quality image should be forever in the news feed (replace 4K image).
- [x] Creating a post and then immediately switch apps & coming back leads to a "failed to create post". Tapping to retry works, but then creates another post. We should auto retry at least a few times if possible before showing the failed msg. 
- [ ] Pressing “post” to submit a comment should close the comments window. The new comment should be instantly visible below the post in the news feed.
- [x] I had a post which was uploaded, I switched apps, came back, then it was forever stuck at 0%. Then I tapped the picture, and the app closed and re opened (no crash), then the post was gone forever. 

# To be done
- [ ] Get BETA back up & running (on test flight). 
- [ ] Add image metadata headers to post upload (EXIF, IPTC, JFIF, TIFF)
- [ ] Album name should be visible while album is open 
- [x] Glow (color border) around profile picture at the top of the news feed should disappear once you have viewed all of that persons stories. It should glow again if the person posts a new story. 
- [ ] Add failed verification to upload status window (instead of saying Done, when it failed verification). Passing verification should continue to say "Done". 
- [ ] Add blur & "Show Anyway" button over posts which have failed verification. This should be visible in the news feed & a user profile. Both the user who created the post, and followers should see this button over the image the first time they see it. The image should instantly appear after tapping the button (should already be cached in the background, just not visible). 
- [x] Add text below each navigation bar icon. "Home, Discover, + (No text), Dating, Me" See Tik Tok for design inspiration.
- [ ] Users should not have to sign in again after sign up with email & password. This will cause a big drop off in completed registrations. We should hit the "login" button automatically for them after a completed signup. 
- [x] add onboarding tutorial which takes them to the discover page to find new content. This should happen after the user is shown how to scroll up/down. 
- [ ] Always Add "You're all caught up" to news feed. Currently, it disappears after a refresh. 
- [x] Add Facebook SDK to app & register it so Facebook Ads can detect app installs for marketing. Please slack me for the facebook developers account login credentials. https://developers.facebook.com/docs/app-ads/app-setup/
- [x] Add "You're All Caught Up" to news feed once the user reaches only posts which he/she has viewed (has seen all posts). This should be calculated by checking the `Post.viewedStatus` value in the `User.feed` query. Backend support: https://github.com/real-social-media/backend/issues/22
- [x] Anytime a user views a comment, submit a graphQL "read" mutation, to create a view record for that comment. This is similar to viewing posts. This way the backend can be notified whenever any user views a comment & it knows the total views per comment along with who viewed it.
  Note: there is no way for the frontend to query a comments "views". This is for the backend only. There will be no query to get all views for a comment. **Only a post property `Post.hasNewCommentActivity` (true/false) which indicates if a given post has any comments which are unread by the current user who made the query.** Later the frontend will show an icon on posts which have `Post.hasNewCommentActivity = true` informing the user that there are comments which have not been read on that post.  Backend support: https://github.com/real-social-media/backend/issues/21
- [x] Posts in the news feed which were created by the current user, should have a red alert icon next to "view more" comments if that post has the backend property set `Post.hasNewCommentActivity == true`) Backend support: https://github.com/real-social-media/backend/issues/21
- [x] When viewing your profile, thumbnails of posts which have a comment which has not yet been viewed (post.all_caught_up_comments == false) should show a red bell icon on the top right of that post thumbnail. This should go away if the user views all comments for that post. **The front-end should read a posts `Post.hasNewCommentActivity` (true/false) attribute** If false, show the alert icon on that post, if true, show no alert icon. Backend support: https://github.com/real-social-media/backend/issues/21
- [x] A red alert icon should appear in the top right corner of the user profile icon on the bottom right of the navigation bar if the user ever has any post with `Post.hasNewCommentActivity == false`. The new boolean field `User.postHasNewCommentActivity` provides this. Backend support: https://github.com/real-social-media/backend/issues/21
- [x] Add ability to edit / delete comments. Both the post owner & the comment creator should have this ability.
- [x] Add ability to edit album titles, archive albums, and delete albums
- [ ] Add support for clicking links in text posts, URLs should be detected by the front-end, and show as a blue hyperlink. URLs contain "www", "http", "https", ".com", "//". Tapping a URL should open that website in the local browser on that device. 
- [x] "update" button on the top right of the "Edit Post" menu, should save & close the edit post menu once the update has been submitted. Currently the update button appears to do nothing. The "Save" button should also close the "Edit Post" menu. 
- [ ] Add confirmation to blocking users "Are you sure you want to block #{username}?" (Yes/No)
- [ ] Swipe up on story while viewing story to see list of views & comments 
- [ ] Improve quality of in-app camera. Currently our in-app camera is very poor / low quality when compared to the iOS native camera. It's also zoomed in super far for some reason. 
- [ ] Users should see an interactive on-boarding overlay which explains that they can tap the right half of a post to "skip" to the next post. This should not have any text if possible (to make it easier to work across languages) but it should be interactive. For example, a floating hand icon, demonstrating tapping on the center right of the screen, once the user taps, take them down to the next post, then again (2x), then back up 2x. **An interactive guide with no text.** All users should see this guide (new and old) unless they have already completed it. The already completed flag should be stored in the frontend on the client side (not server). 
- [x] posts which have property set "Original Post" (not NULL) should show under the the username "reposted from origional_username" where origional_username is the username of the original poster.
- [ ] add ability for users to edit their preferred username on the edit profile screen. This should support errors if the username is taken.
- [x] When creating or editing a post, users should be able to tap on (day, week, month, year, forever) on the timeline slider when creating a post to quickly snap. They should also still be able to drag as they currently do. 
- [x] Our app should refresh the news feed each time the app is resumed (minimized & maximized) or made active. Users switching to another app, then back to our app, should see the news feed loading (loading wheel at the top of the news feed). 
- [x] Rename "First Like" under privacy setting of post & under mental health settings.
- [ ] When tagging users in a post description or comment, after typing the "@" symbol, a dropdown should appear with user search results. This will help users tagging other users, from mistyping or mis remembering a username. Most other social platforms have this functionality.
- [x] move "Repost on REAL" to the top of the share menu
- [x] When "reposting on REAL" the post generated **should not have a watermark.** This will allow the backend to save storage by recognizing that it is the same picture & it will allow the original poster to get views from reposts. There should then be a backend post property "original" which the backend will automatically set once it realizes the new post is the same image data as the old post.
- [x] tap to scroll down feels less snappy (more laggy / low FPS). Please explain the calculation being used to jump to the center of the next post. Maybe we can come up with a way to cache some variables in that calculation to improve performance of skipping between posts.
- [x] Tapping "Follow" in the user search doesn’t work. Instead, you must open their profile to follow them.
- [x] New Users who just installed the app for the first time are getting "Error Occured Failed to create  account" on the email verification in the latest build. 
- [x] When tapping the right or left half of a post in the news feed to scroll down, the next image should be centered, even if it is square (smaller). Currently the next image is always scrolled to be at the top of the device. This makes it difficult to skip to the next picture without moving your thumb.
- [x] Disable swipe to open chat while in create post menu. I can’t use timeline slider because of this to set when a post expires.
- [x] Share to instagram shares the wrong (not selected) picture to instagram. It opens instagram trying to instead share a picture I had in my clipboard (copied), but never had in REAL.
- [x] Repost on real works, but should take the user to the create post screen so they can create a caption, lifetime, etc... 
- [x]  Move "Repost on REAL" to the top of the share screen
- [x] Save to photos should have a message once it's completed successfully such as "Successfully Saved to camera roll"
- [x] The REAL watermark should be white (currently it is black) when sharing photos to instagram, camera roll, etc...
- [ ] Test creating an account, making it private, pending follow requests, etc...
- [ ] Add the option to share to real from the iOS photos app.
- [x] pinch & zoom + it should support moving while zoomed & download the native quality after the user zooms, to replace it with the photo mid zoom once the download completes. Pinch & Zoom - Once a user begins to pinch & zoom on an image, the native quality JPEG image (largest filesize) should be downloaded in the background & replaced with the image the user is viewing (zoomed in on) when it completes download. For example, as a user who wants to zoom in on a picture, I pinch & zoom, notice the quality is a bit blurry as I zoom in, but then it clears up after a few seconds while our app is downloading the native quality post in the background. This is a super important feature, as it encourages our users to upload high quality photos (not blurry ones). 
- [x] Rename “Edit Post” button to “Save” 
- [x] Lifetime slider should be visible (not in a drop down) to make it faster for people to make stories 
- [ ] Allow users to rotate images in the "Post Preview" page. Rotating should be a slider, so users can rotate the image any desired amount down to the degree. Rotating an image should keep the original image unmodified in the iOS photo library & not save the rotated image to camera roll, unless the user later decides to download it.
- [x] Post height limit: currently I am able to upload posts which are greater in height than my device height. I have to scroll down to see the bottom of the post. We should scale photos down to be a max of the device height. 
- [x] "+" menu (bottom menu with take a picture, upload from library, or text only post) should close after a post is uploaded successfully. Currently after uploading a post I still see the menu asking me to choose between camera library, text only, etc... 
- [ ] Should be able to close stories by swiping left/right
- [x] Seen by 0 people shouldn’t be visible. Don’t show “seen by” unless total views > 0 
- [x] The news feed does not seem to refresh very frequently. When does it refresh? For example, most times I click "seen by 0 people" it shows 5+ views & updates the news feed.
- [ ] Selecting an album when creating a post is not intuitive if you have more than 2 albums. Users don't know they can scroll left/right on the album slider. If possible, we should really use a dropdown. One of the options in the dropdown should be "Create New Album". **If this is not possible, at the very least, we should limit the albums shown in the current slider to 1 album (the most recent), with "View More" as the 2nd option. This way "View More" is visible. 
- [ ] Add support to drag (rearrange) posts in an album to reorder them. 
- [x] rename "Create Album" in "Albums" to be "New Album +" with a plus sign visible. Also, please move the “Create Album +” button to the top of the page above all existing albums, so the user does not need to scroll to see it. 
- [x] Add support for the poster & commenter to delete his/her comment. Currently it appears nobody can delete comments in 1.2.3.178. Note: This may be because I opened the post by searching for the users profile, tapping it, then tapping the post, then comments. 
- [x] Discuss new designs with me (Ian) for overlaying captions over posts to help make posts & their captions all fit on the device screen at once (without the user having to scroll to see the caption).
- [x] "Expires in a day/month etc..." should be replaced with the following: If the post expires in <24 hours, it should say "Expires in x hours". If the post expires in 24 hours, it should say "Expires in a day". If the post expires in > 24 hours, it should say "Expires in x days".
- [ ] Add delete option to archived posts, currently users can only “restore from archived” but can’t delete until they restore. 
- [ ] Add confirmation to delete post. This popup should appear for deleting posts in the user profile, new feed, and archived posts. It should say "Are you sure you want to delete this post?"
- [x] Create front-end deployment script which updates the translations file & deploys a new translations file with the filename as the MD5 of the file. The new front-end should now point to this filename. (This will allow older versions of the front-end to use older translations files). It would also be great to have the front-end swap key-values for all languages which don't match en. now on when I (Ian), edit the local file, I will "find & replace all" swap the key / value for all languages. Because we will have a versioned file, this shouldn't break older versions of the UI. This will all become super important as we grow to support other languages & multiple people are collaborating on the translations github. 

- [ ] Add Apple Pay (get diamond) button. https://docs.google.com/document/d/185EGRcI6aPHQvjGqN12InVzRZOqp1ofVkCswswGcfwg/edit?usp=sharing

- [ ] Restrict diamond only features to diamond members only https://docs.google.com/document/d/1JpUly9gC0zZirDC7-Ny0HHhmszMDyc5AI7Ikcd1QCOs/edit?usp=sharing

- [ ] (Feature) Profile link (e.g real.app/username) in any web browser should open our app to that user profile. 

- [ ] (Feature) Post link (e.g real.app/username/post_id) in any web browser should open our app to that user profile.

- [ ] (Feature) Add "Copy Link" to share post options. It should copy the post URL (above) to the iOS clipboard.

- [ ] (Feature) Add "Share Link" to share post options. Tapping this should open the native iOS menu with share options for sharing text (link).

- [ ] Swiping up in camera (while the camera is open) should also open photo library to import a photo 

- [ ] Photo uploads should be retried & retries should exponentially backoff until they give up at 30 minutes. After 30 minutes, all posts which failed to upload should have a failed icon over them on the user profile screen & should show the user a tap "retry upload" button. The user should also see an error that the post failed on their home screen. Other posts should continue uploading in parallel so this post does not stop them. 

- [ ] Followers list seems to be cut off, a bunch of usernames are missing which follow content@real.app for example, doesn't show me as a follower, but I follow that account.

- [ ] Camera should open to selfie mode (mirrored), as instagram's camera does.

- [ ] Bug: When taking a picture using the selfie camera, it's zoomed in for some reason. This does not happen in other apps with in-app cameras.

- [ ] post sharing (backend **per post** public property) - When disabled, the frontend should hide the share button on this post from followers who see the post in their news feed. This should be enabled by default. This setting should be visible each time a user creates a post (if they want anyone who sees it to be able to share it). This should have subtext "Allows other users to share your posts).

- [ ] Post Sharing (Mental Health Setting). Setting to enabled, shows the option for post sharing (defaulted to true) when creating a post. Setting it to disabled, removes that option (or greys it out) when creating a post. This should have subtext "Allows other users to share your posts).
- [ ] Add iPad support (responsive, portrait only).

- [ ] Deploy app with iPad supported.

- [ ] Add facebook login/signup

- [ ] Add mental health dropdown option "Daily Post Limit" with dropdown options: "Unlimited" default, "10", "20"..."100". Description (heading 2 text under the label/dropdown): "Limit the total number of posts you can see per day.".

- [ ] ZenDesk Support https://github.com/Imcloug/REAL-FrontEnd/issues/95

- [ ] Add "tap to focus on area" to in-app camera. 

- [ ] Add "pinch & zoom" to in-app camera.

- [ ] Implement allowing the users to transfer their REAL account balance to their PayPal/Braintree account via the PayPal payouts https://developer.paypal.com/docs/payouts/ 

- [ ] Typing comments which are longer than the text field doesn’t work. Please limit comments to the caption length but allow the text field to scroll.
- [ ] The camera & chat icons are visible on the login screen which is confusing.
- [ ] Apply for Diamond Button - https://github.com/Imcloug/REAL-FrontEnd/issues/99
- [ ] Finish Real Dating MVP (will add details)

- [ ] Android support.
